### PR TITLE
Initialise: make snapshots instead of prepending state events

### PR DESCRIPTION
**Pull request by DMRobertson**
Primarily a perf fix for when the proxy re-joins a room, but it's also arguably a correctness fix because we have no idea how these things appear in the timeline.

### CODE REVIEW

**Comment by DMRobertson** [comment]
part of #232 #294

File: state/accumulator.go [hunk 1]
```diff
@@ -650,3 +630,82 @@ func (a *Accumulator) filterToNewTimelineEvents(txn *sqlx.Tx, dedupedEvents []Ev
...
+		return internal.NewDataError(errMsg)
+	}
+	return nil
+}
+
+type stateMap struct {
+	// state_key (user id) -> NID
+	Memberships map[string]int64
+	// type, state_key -> NID
+	Other map[[2]string]int64
```
Comment (DMRobertson) [inline state/accumulator.go:665]: Not sure if this was worth splitting up, but thinking was that there's no point storing lots of copies of the string "`m.room.member"` in large rooms.

File: state/accumulator.go [hunk 2]
```diff
@@ -650,3 +630,82 @@ func (a *Accumulator) filterToNewTimelineEvents(txn *sqlx.Tx, dedupedEvents []Ev
...
+	if e.Type == "m.room.member" {
+		replacedNID = s.Memberships[e.StateKey]
+		s.Memberships[e.StateKey] = e.NID
+	} else {
+		key := [2]string{e.Type, e.StateKey}
+		replacedNID = s.Other[key]
+		s.Other[key] = e.NID
+	}
+	return
+}
```
Comment (DMRobertson) [inline state/accumulator.go:678]: I don't actually use the return value here. At some point I was more ambitious and wanted to reuse this machinery in Accumulate... but then I decided that was not really getting us anything and left it for a rainy day.

File: state/accumulator.go [hunk 3]
```diff
@@ -650,3 +630,82 @@ func (a *Accumulator) filterToNewTimelineEvents(txn *sqlx.Tx, dedupedEvents []Ev
...
+	membershipNIDs = make([]int64, 0, len(s.Memberships))
+	otherNIDs = make([]int64, 0, len(s.Other))
+	for _, nid := range s.Memberships {
+		membershipNIDs = append(membershipNIDs, nid)
+	}
+	for _, nid := range s.Other {
+		otherNIDs = append(otherNIDs, nid)
+	}
+	return
+}
```
Comment (DMRobertson) [inline state/accumulator.go:690]: It'd a bit of a shame to convert from []Event to a map and then back to []Event again if we're only changing the state by one event. But we're optimising for leaving and rejoining large rooms here, with large gappy state blocks. We've seen gappy state blocks of size multiple thousands on the m.org deployment!

File: state/accumulator.go [hunk 4]
```diff
@@ -650,3 +630,82 @@ func (a *Accumulator) filterToNewTimelineEvents(txn *sqlx.Tx, dedupedEvents []Ev
...
+
+func (a *Accumulator) stateMapAtSnapshot(txn *sqlx.Tx, snapID int64) (stateMap, error) {
+	snapshot, err := a.snapshotTable.Select(txn, snapID)
+	if err != nil {
+		return stateMap{}, err
+	}
+	// pull stripped events as this may be huge (think Matrix HQ)
+	events, err := a.eventsTable.SelectStrippedEventsByNIDs(txn, true, append(snapshot.MembershipEvents, snapshot.OtherEvents...))
+	if err != nil {
+		return stateMap{}, err
```
Comment (DMRobertson) [inline state/accumulator.go:700]: We only want event NID, type and state key here, so we could use a more specialised query here to fetch less data.

File: tests-e2e/gappy_state_test.go [hunk 5]
```diff
@@ -71,28 +79,50 @@ func TestGappyState(t *testing.T) {
...
+
+	t.Log("Alice requests an initial sliding sync on device 2, with timeline limit big enough to see her first message at the start of the test.")
 	syncResp = alice.SlidingSync(t,
 		sync3.Request{
 			Lists: map[string]sync3.RequestList{
 				"a": {
 					Ranges: [][2]int64{{0, 20}},
 					RoomSubscription: sync3.RoomSubscription{
-						TimelineLimit: 10,
+						TimelineLimit: 100,
```
Comment (DMRobertson) [inline tests-e2e/gappy_state_test.go:91]: This test does produce a limited sync for the poller. But that poll is for a new device so gets timelimit limit 1 which wasn't satisfactory---I wanted to confirm that we include timeline events after the gap, but not before.

I felt it was easier to write an integration test so I can use the same poller without any logging out and in shenanigans.

File: state/accumulator.go [hunk 6]
```diff
@@ -149,19 +149,33 @@ type InitialiseResult struct {
...
+//  2. Fetch the current state of the room, as a map from (type, state_key) to event.
+//     If there is no existing state snapshot, this map is the empty map.
+//
+//     If the state hasn't altered, bail.
+//
+//  4. Create new snapshot.
+//      Update the map from (2) with the events in  in `state`.  (There must be similar logic already in A ccumulate for this?)
+//      Store the snapshot. Mark the room's current state as being this snapshot.
+//
+// 5. If the starting snapshot ID was not zero, emit a cache invalidation payload.
```
[the remark this answers is not in the document]
Comment (kegsay) [reply state/accumulator.go:177]: Those tests are only testing cache invalidation via redaction, not cache invalidation via joining a room after a long period of no one being in the room. The latter case is what we're trying to catch here.

[the remark this answers is not in the document]
Comment (DMRobertson) [reply state/accumulator.go:177]: I don't follow. I specifically mean TestGappyState and `TestGappyStateDoesNotAccumulateTheStateBlock`, neither of which emit a redaction event.

File: state/accumulator.go [hunk 7]
```diff
@@ -149,19 +149,33 @@ type InitialiseResult struct {
...
+//  0. Ensure the state block is not empty.
 //
-// and otherwise does nothing.
+//  1. Capture the current snapshot ID, possibly zero. If it is zero, ensure that the
+//      state block contains a `create event`.
+//
+//  2. Insert the events and determine if the state block alters the current state of
+//     the room from the proxy's POV. This is the case iff some event in the state block
+//     is unknown to the proxy. (We choose to ignore the possibility that a poller is
+//     slow and gives us e.g. a subset of current room state.)
```
[the remark this answers is not in the document]
Comment (DMRobertson) [reply state/accumulator.go:166]: Thanks, will use this wording.

File: state/accumulator.go [hunk 8]
```diff
@@ -213,71 +199,65 @@ func (a *Accumulator) Initialise(roomID string, state []json.RawMessage) (Initia
...
+			currentState.Ingest(ev)
+		}
+		memberNIDs, otherNIDs := currentState.NIDs()
 		snapshot := &SnapshotRow{
 			RoomID:           roomID,
-			MembershipEvents: pq.Int64Array(memberNIDs),
-			OtherEvents:      pq.Int64Array(otherNIDs),
+			MembershipEvents: memberNIDs,
+			OtherEvents:      otherNIDs,
 		}
```
[the remark this answers is not in the document]
Comment (DMRobertson) [reply state/accumulator.go:263]: Oh I see, the caller needs to do this. That means 

> // 5. If the starting snapshot ID was not zero, emit a cache invalidation payload.

is misleading.

[the remark this answers is not in the document]
Comment (DMRobertson) [reply state/accumulator.go:263]: I'll probably change this to a ReplacedSnapshot field, true iff the caller should send an invalidation payload.

**Comment by DMRobertson** [comment]
I had to rethink a lot of this. -> #329.

**Pull request closed without merging**